src/Nethermind/Nethermind.Specs/ChainSpecStyle/ChainSpecParametersProvider.cs

```diff
@@ -0,0 +1,99 @@
+// SPDX-FileCopyrightText: 2024 Demerzel Solutions Limited
+// SPDX-License-Identifier: LGPL-3.0-only
+
+
+using System.Reflection;
+using System.Text.Json;
+using Nethermind.Config;
+
+namespace Nethermind.Specs.ChainSpecStyle;
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using Nethermind.Core;
+
+public class ChainSpecParametersProvider : IChainSpecParametersProvider
+{
+    // TODO: test that all IChainSpecEngineParameters have this suffix
+    private const string EngineParamsSuffix = "ChainSpecEngineParameters";
+
+    private readonly Dictionary<string, JsonElement> _chainSpecParameters =
+        new(StringComparer.InvariantCultureIgnoreCase);
+
+    private readonly ConcurrentDictionary<Type, IChainSpecEngineParameters> _instances = new();
+    public string SealEngineType { get; }
+
+    public ChainSpecParametersProvider(Dictionary<string, JsonElement> engineParameters)
+    {
+        foreach (var parameters in engineParameters)
+        {
+            _chainSpecParameters[parameters.Key] = parameters.Value;
+        }
+
+        InitializeInstances();
+        SealEngineType = CalculateSealEngineType();
+    }
+
+    string CalculateSealEngineType()
+    {
+        string? result = null;
+        foreach (IChainSpecEngineParameters item in _instances.Values)
+        {
+            if (item.SealEngineType is not null)
+            {
+                if (result is not null)
+                {
+                    throw new InvalidOperationException("Multiple seal engines in chain spec");
+                }
+
+                result = item.SealEngineType;
+            }
+        }
+
+        if (result is null)
+        {
+            throw new InvalidOperationException("No seal engine in chain spec");
+        }
+
+        return result;
+    }
+
+    private void InitializeInstances()
+    {
+        Type type = typeof(IChainSpecEngineParameters);
+        IEnumerable<Type> types = TypeDiscovery.FindNethermindBasedTypes(type).Where(x => x.IsClass);
+        foreach (Type @class in types)
+        {
+            string engineName = @class.Name.Remove(@class.Name.Length - EngineParamsSuffix.Length);
```

Comment:
Not sure what you mean. It already looks for all inheritors of `IChainSpecEngineParameters`. I didn't make manual registration because ChainSpec is more similar to `IConfig`. So I thought we should behave the same way here

Reply:
My question was more in line with replacing the class name logic with something more "typesafe": right now it seems like I need to follow a specific naming convention but it's not enforced statically, and to figure it out I need to take a look into this implementation.

If I were to add another `IChainSpecEngineParameters` class, what would the requirements be and how would I figure them out?

Reply:
I think what Lautaro means is that we should have a property `IChainSpecEngineParamerters.Name` or something similar, instead of using the name of the class. I understand @deffrian's point that we're being consistent with how `IConfig` works. But maybe we should start improving both.

Reply:
> I think what Lautaro means is that we should have a property `IChainSpecEngineParamerters.Name` or something similar, instead of using the name of the class.

Precisely, I just didn't want to push for any particular direction.

> we're being consistent with how `IConfig` works

Didn't know that we were using the class name approach there. If it is consistent with the existing style then let's move forward, but we might want to add a TODO to eventually improve (that is, move to the type system) these kind of requirements.